File: sorting/sort_colors.py
class Solution:
    def sortColors(self, nums: list[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        # dutch national flag problem
        
        l = 0
        r = len(nums) - 1
        i = 0

        while i <= r:
            if nums[i] == 0:
                nums[i], nums[l] = nums[l], nums[i]
                l += 1
                i += 1
            elif nums[i] == 2:
                nums[i], nums[r] = nums[r], nums[i]
                r -= 1
            else:
                i += 1

s = Solution()
s.sortColors([2,0,2,1,1,0])

File: leet/sorting/sort_colors.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    void sortColors(vector<int>& nums) {
        // 2 0 1
        // 1 0 2
        // classic dutch flag
        int n = nums.size();
        int l = 0;
        int r = n-1;
        int i = 0;

        for (int i = 0; i <= r;) {
            if (nums[i] == 0) {
                swap(nums[i], nums[l]);
                l++;
                i++;
            } else if (nums[i] == 2) {
                swap(nums[i], nums[r]);
                r--;
            } else {
                i++;
            }
        }
    }
};